File: HPWorkpath_v1.6.3/Samples/ExampleAPIServices/source/ScanSample/src/com/hp/workpath/sample/scansample/fragments/MarginsPreference.java
// Copyright 2018 HP Development Company, L.P.
// SPDX-License-Identifier: MIT
package com.hp.workpath.sample.scansample.fragments;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Handler;
import android.util.AttributeSet;

import androidx.preference.DialogPreference;
import androidx.preference.PreferenceManager;
import androidx.preference.PreferenceViewHolder;

import com.hp.workpath.sample.scansample.R;

public class MarginsPreference extends DialogPreference {


    float mLeft = 0.0f;
    float mTop = 0.0f;
    float mRight = 0.0f;
    float mBottom = 0.0f;

    float mLeftMin = 0.0f;
    float mTopMin = 0.0f;
    float mRightMin = 0.0f;
    float mBottomMin = 0.0f;

    float mLeftMax = 0.0f;
    float mTopMax = 0.0f;
    float mRightMax = 0.0f;
    float mBottomMax = 0.0f;

    public MarginsPreference(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    public MarginsPreference(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public MarginsPreference(Context context) {
        super(context);
    }

    @Override
    public void onBindViewHolder(PreferenceViewHolder view) {
        super.onBindViewHolder(view);
        setDialogLayoutResource(R.layout.layout_margins);
        loadMargins();
    }

    public void applyMargins() throws Exception {
        if (!isInRange(mLeftMin, mLeftMax, mLeft)) {
            throw new Exception(getContext().getString(R.string.range_margin, mLeftMin, mLeftMax));
        }
        if (!isInRange(mTopMin, mTopMax, mTop)) {
            throw new Exception(getContext().getString(R.string.range_margin, mTopMin, mTopMax));
        }
        if (!isInRange(mRightMin, mRightMax, mRight)) {
            throw new Exception(getContext().getString(R.string.range_margin, mRightMin, mRightMax));
        }
        if (!isInRange(mBottomMin, mBottomMax, mBottom)) {
            throw new Exception(getContext().getString(R.string.range_margin, mBottomMin, mBottomMax));
        }
        saveMargins();
    }

    private void loadMargins() {
        SharedPreferences sharedPref = PreferenceManager.getDefaultSharedPreferences(getContext());
        mLeft = sharedPref.getFloat(getKey() + "Left", 0.0f);
        mTop = sharedPref.getFloat(getKey() + "Top", 0.0f);
        mRight = sharedPref.getFloat(getKey() + "Right", 0.0f);
        mBottom = sharedPref.getFloat(getKey() + "Bottom", 0.0f);
        setMarginSummary();
    }

    private void setMarginSummary() {
        Handler handler = new Handler();
        final Runnable r = () -> setSummary(getContext().getString(R.string.summary_margin, mLeft, mTop, mRight, mBottom));
        handler.post(r);
    }

    private void saveMargins() {
        SharedPreferences sharedPref = PreferenceManager.getDefaultSharedPreferences(getContext());
        SharedPreferences.Editor editor = sharedPref.edit();
        editor.putFloat(getKey() + "Left", mLeft);
        editor.putFloat(getKey() + "Top", mTop);
        editor.putFloat(getKey() + "Right", mRight);
        editor.putFloat(getKey() + "Bottom", mBottom);
        editor.apply();
        setMarginSummary();
    }

    public void setLeftLimits(final float min, final float max) {
        mLeftMin = min;
        mLeftMax = max;
    }

    public void setTopLimits(final float min, final float max) {
        mTopMin = min;
        mTopMax = max;
    }

    public void setRightLimits(final float min, final float max) {
        mRightMin = min;
        mRightMax = max;
    }

    public void setBottomLimits(final float min, final float max) {
        mBottomMin = min;
        mBottomMax = max;
    }

    public void setMargins(float left, float top, float right, float bottom) {
        this.mLeft = left;
        this.mTop = top;
        this.mRight = right;
        this.mBottom = bottom;
    }

    public float getLeft() {
        return mLeft;
    }

    public float getTop() {
        return mTop;
    }

    public float getRight() {
        return mRight;
    }

    public float getBottom() {
        return mBottom;
    }

    private boolean isInRange(final float min, final float max, final float value) {
        return max > min ? value >= min && value <= max : value >= max && value <= min;
    }
}

File: HPWorkpath_v1.6.3/Samples_Kotlin/ExampleAPIServices/source/ScanSample/src/com/hp/workpath/sample/scansample/fragments/MarginsPreference.kt
// Copyright 2018 HP Development Company, L.P.
// SPDX-License-Identifier: MIT
package com.hp.workpath.sample.scansample.fragments

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.AttributeSet
import android.util.Log
import androidx.preference.DialogPreference
import androidx.preference.PreferenceManager
import androidx.preference.PreferenceViewHolder
import com.hp.workpath.sample.scansample.R

class MarginsPreference : DialogPreference {
    var left = 0.0f
    var top = 0.0f
    var right = 0.0f
    var bottom = 0.0f
    private var mLeftMin = 0.0f
    private var mTopMin = 0.0f
    private var mRightMin = 0.0f
    private var mBottomMin = 0.0f
    private var mLeftMax = 0.0f
    private var mTopMax = 0.0f
    private var mRightMax = 0.0f
    private var mBottomMax = 0.0f

    constructor(context: Context?, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {}
    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs) {}
    constructor(context: Context?) : super(context) {}

    override fun onBindViewHolder(view: PreferenceViewHolder) {
        super.onBindViewHolder(view)
        dialogLayoutResource = R.layout.layout_margins
        loadMargins()
    }

    @Throws(Throwable::class)
    fun applyMargins() {
        if (left !in mLeftMin..mLeftMax) {
            throw Throwable(context.getString(R.string.range_margin, mLeftMin, mLeftMax))
        }
        if (top !in mTopMin..mTopMax) {
            throw Throwable(context.getString(R.string.range_margin, mTopMin, mTopMax))
        }
        if (right !in mRightMin..mRightMax) {
            throw Throwable(context.getString(R.string.range_margin, mRightMin, mRightMax))
        }
        if (bottom !in mBottomMin..mBottomMax) {
            throw Throwable(context.getString(R.string.range_margin, mBottomMin, mBottomMax))
        }
        saveMargins()
    }

    private fun loadMargins() {
        val sharedPref = PreferenceManager.getDefaultSharedPreferences(context)
        left = sharedPref.getFloat(key + "Left", 0.0f)
        top = sharedPref.getFloat(key + "Top", 0.0f)
        right = sharedPref.getFloat(key + "Right", 0.0f)
        bottom = sharedPref.getFloat(key + "Bottom", 0.0f)
        setMarginSummary()
    }

    private fun setMarginSummary() {
        val handler = Handler(Looper.getMainLooper())
        val r = Runnable { summary = context.getString(R.string.summary_margin, left, top, right, bottom) }
        handler.post(r)
    }

    private fun saveMargins() {
        val sharedPref = PreferenceManager.getDefaultSharedPreferences(context)
        val editor = sharedPref.edit()
        editor.putFloat(key + "Left", left)
        editor.putFloat(key + "Top", top)
        editor.putFloat(key + "Right", right)
        editor.putFloat(key + "Bottom", bottom)
        editor.apply()
        setMarginSummary()
    }

    fun setLeftLimits(min: Float, max: Float) {
        mLeftMin = min
        mLeftMax = max
    }

    fun setTopLimits(min: Float, max: Float) {
        mTopMin = min
        mTopMax = max
    }

    fun setRightLimits(min: Float, max: Float) {
        mRightMin = min
        mRightMax = max
    }

    fun setBottomLimits(min: Float, max: Float) {
        mBottomMin = min
        mBottomMax = max
    }

    fun setMargins(left: Float, top: Float, right: Float, bottom: Float) {
        this.left = left
        this.top = top
        this.right = right
        this.bottom = bottom
    }
}